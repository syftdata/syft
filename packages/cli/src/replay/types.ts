export interface TSSourcePath {
  fileName: string;
}
export interface EventSource {
  name: string;
  uiEventType: string;
  source: TSSourcePath;
  owner?: string;
  ownerSource?: TSSourcePath;
}
