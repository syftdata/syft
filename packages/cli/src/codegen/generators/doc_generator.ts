import * as fs from 'fs';
import { logInfo } from '../../utils';
import { type AST } from '@syftdata/common/lib/types';

export function generate(ast: AST, destDir: string): void {
  fs.mkdirSync(destDir, { recursive: true });
  logInfo('This functionality is not yet available :sad:');
  // now write one .md file per event. under a source directory.
}
