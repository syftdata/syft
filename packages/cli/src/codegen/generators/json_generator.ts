import * as fs from 'fs';
import * as path from 'path';
import { type AST } from '@syftdata/common/lib/types';

function replacer(key: string, value: any): any {
  if (value != null && typeof value === 'object' && !Array.isArray(value)) {
    const replacement = {};
    Object.keys(value).forEach((k) => {
      if (k === 'syftConfig') return;
      replacement[k] = value[k];
    });
    return replacement;
  }
  return value;
}

export function generate(ast: AST, destDir: string): void {
  fs.mkdirSync(destDir, { recursive: true });
  ast.eventSchemas = ast.eventSchemas.filter((schema) => {
    return schema.exported === true;
  });
  const astJson = JSON.stringify(ast, replacer, 2);
  fs.writeFileSync(path.join(destDir, 'schema.json'), astJson);
}
