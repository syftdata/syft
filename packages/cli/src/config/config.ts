import type { StaticConfig } from '../client_types';
import type { ObjectLiteralExpression, Project } from 'ts-morph';
import { SyntaxKind, PropertyAssignment } from 'ts-morph';
import { createTSProject } from '../codegen/compiler';

const STATIC_CONFIG_TYPE = 'StaticConfig';

export function getConfigExpressionForProject(
  project: Project
): ObjectLiteralExpression | undefined {
  let syftConfig: ObjectLiteralExpression | undefined;
  for (const sourceFile of project.getSourceFiles()) {
    sourceFile.getVariableDeclarations().forEach((variable) => {
      let varialbeType = variable.getType().getText();
      const symbol = variable.getType().getSymbol();
      if (symbol !== undefined) {
        varialbeType = symbol.getEscapedName();
      }
      if (varialbeType !== STATIC_CONFIG_TYPE) return;
      syftConfig = variable.getInitializerIfKind(
        SyntaxKind.ObjectLiteralExpression
      );
    });
    if (syftConfig !== undefined) break;
  }
  return syftConfig;
}

export function getConfigFromExpression(
  expression: ObjectLiteralExpression
): StaticConfig {
  const val: StaticConfig = {
    projectName: 'test',
    version: '0.0.0'
  };
  expression.getProperties().forEach((property) => {
    if (property instanceof PropertyAssignment) {
      const propertyName = property.getSymbolOrThrow().getEscapedName();
      const value = property
        .getInitializerOrThrow()
        .getText()
        .replace(/['"]/g, '');
      if (propertyName === 'version') {
        val[propertyName] = value;
      }
    }
  });
  return val;
}

export function setConfigVersionOnExpression(
  expression: ObjectLiteralExpression,
  version: string
): boolean {
  expression.getProperties().forEach((property) => {
    if (property instanceof PropertyAssignment) {
      const propertyName = property.getSymbolOrThrow().getEscapedName();
      if (propertyName === 'version') {
        property.setInitializer(`'${version}'`);
        return true;
      }
    }
  });
  return false;
}

export function getConfig(schemaFolder: string): StaticConfig | undefined {
  const project = createTSProject([schemaFolder]);
  const expression = getConfigExpressionForProject(project);
  if (expression === undefined) {
    return;
  }
  return getConfigFromExpression(expression);
}
