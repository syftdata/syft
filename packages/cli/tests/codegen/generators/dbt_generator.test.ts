/* eslint-disable @typescript-eslint/no-unused-vars */
import { SyftEventType } from '@syftdata/client';
import * as mockfs from 'mock-fs';
import * as fs from 'fs';
import * as path from 'path';
import { generate } from '../../../src/codegen/generators/dbt_generator';
import { type Field, type AST } from '@syftdata/common/lib/types';
import { type ProviderConfig } from '../../../src/config/sink_configs';
import { getZodTypeForSchema } from '@syftdata/codehandler';

const TEST_FIELDS: Field[] = [
  {
    name: 'backgrounded',
    type: {
      name: 'boolean',
      zodType: 'z.boolean()',
      isArray: false
    },
    isOptional: false
  }
];
const TEST_ZOD_TYPE: string = getZodTypeForSchema(TEST_FIELDS);

afterEach(() => {
  mockfs.restore();
});

describe('generate', () => {
  it('with dbt parses an empty schema file', async () => {
    mockfs();
    const ast: AST = {
      eventSchemas: [],
      config: {
        projectName: 'test',
        version: '1.0.0'
      },
      sinks: [
        {
          id: 'testSink',
          type: 'bigquery',
          config: {
            dataset: 'testDataset',
            projectId: 'testProject'
          }
        }
      ],
      inputs: []
    };
    const providerConfig: ProviderConfig = {
      sdkType: 'Segment'
    };
    generate(ast, 'dbt', providerConfig);
    // now read project file and models from fs.
    expect(
      fs.readFileSync(path.join('dbt', 'testSink', 'dbt_project.yml'), 'utf8')
    ).toMatchSnapshot();
    expect(
      fs.readFileSync(path.join('dbt', 'testSink', 'profiles.yml'), 'utf8')
    ).toMatchSnapshot();
    // creates a model directory with sources and staging models.
    // expect(
    //   fs.readFileSync(
    //     path.join('dbt', 'testSink', 'models', 'schema.yml'),
    //     'utf8'
    //   )
    // ).toMatchSnapshot();
    expect(
      fs.readFileSync(
        path.join('dbt', 'testSink', 'models', 'syft.yml'),
        'utf8'
      )
    ).toMatchSnapshot();
    // expect(
    //   fs.readdirSync(path.join('dbt', 'testSink', 'models')).length
    // ).toEqual(2);
  });

  it('with dbt parses an event with Segment', async () => {
    mockfs();
    const ast: AST = {
      eventSchemas: [
        {
          name: 'TestEvent',
          fields: TEST_FIELDS,
          eventType: SyftEventType.PAGE,
          exported: true,
          zodType: TEST_ZOD_TYPE
        }
      ],
      config: {
        projectName: 'test',
        version: '1.0.0'
      },
      sinks: [
        {
          id: 'testSink',
          type: 'bigquery',
          config: {
            dataset: 'testDataset',
            projectId: 'testProject'
          }
        }
      ],
      inputs: []
    };
    const providerConfig: ProviderConfig = {
      sdkType: 'segment'
    };
    generate(ast, 'dbt', providerConfig);
    // now read project file and models from fs.
    expect(
      fs.readFileSync(path.join('dbt', 'testSink', 'dbt_project.yml'), 'utf8')
    ).toMatchSnapshot();
    expect(
      fs.readFileSync(path.join('dbt', 'testSink', 'profiles.yml'), 'utf8')
    ).toMatchSnapshot();
    // expect(
    //   fs.readdirSync(path.join('dbt', 'testSink', 'models')).length
    // ).toEqual(3);
    // expect(
    //   fs.readFileSync(
    //     path.join('dbt', 'testSink', 'models', 'stg_test_event.sql'),
    //     'utf8'
    //   )
    // ).toMatchSnapshot();
    // expect(
    //   fs.readFileSync(
    //     path.join('dbt', 'testSink', 'models', 'schema.yml'),
    //     'utf8'
    //   )
    // ).toMatchSnapshot();
    expect(
      fs.readFileSync(
        path.join('dbt', 'testSink', 'models', 'syft.yml'),
        'utf8'
      )
    ).toMatchSnapshot();
  });

  it('with dbt parses an event with HeapIO', async () => {
    mockfs();
    const ast: AST = {
      eventSchemas: [
        {
          name: 'TestEvent',
          fields: TEST_FIELDS,
          eventType: SyftEventType.PAGE,
          exported: true,
          zodType: TEST_ZOD_TYPE
        }
      ],
      config: {
        projectName: 'test',
        version: '1.0.0'
      },
      sinks: [
        {
          id: 'testSink',
          type: 'bigquery',
          config: {
            dataset: 'testDataset',
            projectId: 'testProject'
          }
        }
      ],
      inputs: []
    };
    const providerConfig: ProviderConfig = {
      sdkType: 'heap',
      platform: 'Web'
    };
    generate(ast, 'dbt', providerConfig);
    // now read project file and models from fs.
    expect(
      fs.readFileSync(path.join('dbt', 'testSink', 'dbt_project.yml'), 'utf8')
    ).toMatchSnapshot();
    expect(
      fs.readFileSync(path.join('dbt', 'testSink', 'profiles.yml'), 'utf8')
    ).toMatchSnapshot();
    // expect(
    //   fs.readdirSync(path.join('dbt', 'testSink', 'models')).length
    // ).toEqual(3);
    // expect(
    //   fs.readFileSync(
    //     path.join('dbt', 'testSink', 'models', 'stg_test_event.sql'),
    //     'utf8'
    //   )
    // ).toMatchSnapshot();
    // expect(
    //   fs.readFileSync(
    //     path.join('dbt', 'testSink', 'models', 'schema.yml'),
    //     'utf8'
    //   )
    // ).toMatchSnapshot();
    expect(
      fs.readFileSync(
        path.join('dbt', 'testSink', 'models', 'syft.yml'),
        'utf8'
      )
    ).toMatchSnapshot();
  });
  it('with dbt parses an event with Syft', async () => {
    mockfs();
    const ast: AST = {
      eventSchemas: [
        {
          name: 'TestEvent',
          fields: TEST_FIELDS,
          eventType: SyftEventType.PAGE,
          exported: true,
          zodType: TEST_ZOD_TYPE
        }
      ],
      config: {
        projectName: 'test',
        version: '1.0.0'
      },
      sinks: [
        {
          id: 'testSink',
          type: 'bigquery',
          config: {
            dataset: 'testDataset',
            projectId: 'testProject'
          }
        }
      ],
      inputs: []
    };
    const providerConfig: ProviderConfig = {
      sdkType: 'syft'
    };
    generate(ast, 'dbt', providerConfig);
    // now read project file and models from fs.
    expect(
      fs.readFileSync(path.join('dbt', 'testSink', 'dbt_project.yml'), 'utf8')
    ).toMatchSnapshot();
    expect(
      fs.readFileSync(path.join('dbt', 'testSink', 'profiles.yml'), 'utf8')
    ).toMatchSnapshot();
    // expect(
    //   fs.readdirSync(path.join('dbt', 'testSink', 'models')).length
    // ).toEqual(3);
    // expect(
    //   fs.readFileSync(
    //     path.join('dbt', 'testSink', 'models', 'stg_test_event.sql'),
    //     'utf8'
    //   )
    // ).toMatchSnapshot();
    // expect(
    //   fs.readFileSync(
    //     path.join('dbt', 'testSink', 'models', 'schema.yml'),
    //     'utf8'
    //   )
    // ).toMatchSnapshot();
    expect(
      fs.readFileSync(
        path.join('dbt', 'testSink', 'models', 'syft.yml'),
        'utf8'
      )
    ).toMatchSnapshot();
  });
});
