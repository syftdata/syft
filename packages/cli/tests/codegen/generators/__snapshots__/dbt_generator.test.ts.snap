// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate with dbt parses an empty schema file 1`] = `
"name: test
version: 1.0.0
config-version: 2
profile: default
model-paths:
  - models
test-paths:
  - tests
seed-paths:
  - seeds
target-path: target
clean-targets:
  - target
  - dbt_packages
"
`;

exports[`generate with dbt parses an empty schema file 2`] = `
"default:
  target: dev
  outputs:
    dev:
      type: bigquery
      method: oauth
      project: testProject
      dataset: testDataset
      timeout_seconds: 300
"
`;

exports[`generate with dbt parses an empty schema file 3`] = `
"version: 2
sources:
  - schema: testProject
    database: testDataset
    name: testSink
    description: Loaded from Syft
    tables: []
"
`;

exports[`generate with dbt parses an event with HeapIO 1`] = `
"name: test
version: 1.0.0
config-version: 2
profile: default
model-paths:
  - models
test-paths:
  - tests
seed-paths:
  - seeds
target-path: target
clean-targets:
  - target
  - dbt_packages
"
`;

exports[`generate with dbt parses an event with HeapIO 2`] = `
"default:
  target: dev
  outputs:
    dev:
      type: bigquery
      method: oauth
      project: testProject
      dataset: testDataset
      timeout_seconds: 300
"
`;

exports[`generate with dbt parses an event with HeapIO 3`] = `
"version: 2
sources:
  - schema: testProject
    database: testDataset
    name: testSink
    description: Loaded from Syft
    tables:
      - name: test_event
        columns:
          - name: backgrounded
            meta:
              type: boolean
            quote: false
            tests:
              - not_null
          - name: _id
            description: The unique ID of the track call itself.
            meta:
              type: number
              metabase.semantic_type: type/PK
            quote: false
            tests:
              - not_null
          - name: anonymous_id
            description: The anonymous ID of the user.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: context_ip
            description: >-
              The IP address of the client. Non-user-related context fields sent
              with each track call.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: context_library_name
            description: >-
              The Logging library name. Non-user-related context fields sent
              with each track call.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: context_library_version
            description: >-
              The Logging library version. Non-user-related context fields sent
              with each track call.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: context_page_path
            description: The path of the page on which the event was logged.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: context_page_title
            description: The title of the page on which the event was logged.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: context_page_url
            description: The full url of the page on which the event was logged.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: context_locale
            description: The browsers locale used when the event was logged.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: context_user_agent
            description: The browsers user-agent string.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: event
            description: The slug of the event name, so you can join the tracks table.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: event_text
            description: The name of the event.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: loaded_at
            description: >-
              The UTC timestamp reflecting when the data was staged by the
              processor. Bigquery only.
            meta:
              type: timestamp
            quote: false
            tests:
              - not_null
          - name: original_timestamp
            description: >-
              The original timestamp set by the Segment library at the time the
              event is created. This timestamp can be affected by device clock
              skew.
            meta:
              type: timestamp
            quote: false
            tests:
              - not_null
          - name: received_at
            description: When Segment received the track call.
            meta:
              type: timestamp
            quote: false
            tests:
              - not_null
          - name: sent_at
            description: >-
              When a user triggered the track call. This timestamp can also be
              affected by device clock skew
            meta:
              type: timestamp
            quote: false
            tests:
              - not_null
          - name: timestamp
            description: The UTC-converted timestamp which is set by the Segment library
            meta:
              type: timestamp
            quote: false
            tests:
              - not_null
          - name: uuid_ts
            description: >-
              This is used to keep track of when the specific event was last
              processed by our connector, specifically for deduping and
              debugging purposes. You can generally ignore this column
            meta:
              type: timestamp
            quote: false
            tests:
              - not_null
          - name: event_id
            description: Unique ID of the associated event, randomly generated by Heap.
            meta:
              type: number
            quote: false
            tests:
              - not_null
          - name: user_id
            description: Unique ID of the associated user, randomly generated by Heap.
            meta:
              type: number
            quote: false
            tests:
              - not_null
          - name: session_id
            description: Unique ID of the associated session, randomly generated by Heap.
            meta:
              type: number
            quote: false
            tests:
              - not_null
          - name: time
            description: Timestamp without timezone of when the event happened.
            meta:
              type: timestamp
            quote: false
            tests:
              - not_null
          - name: type
            description: >-
              For web autocaptured events, can be any of view page, click,
              submit, change, with push state events registered as view page
              events. For iOS auto-tracked events, can be touch, edit field, or
              a gesture recognizer you’ve defined. For custom events, this will
              be the custom event name.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: library
            description: >-
              Version of Heap library on which event occurred. Can be one of
              “web”, “iOS”, or “server”.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: platform
            description: User’s operating system.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: device_type
            description: Device type, which can be one of “Mobile”, “Tablet”, or “Desktop”.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: country
            description: Country in which user session occurred, based on IP.
            meta:
              type: string
              metabase.semantic_type: type/Country
            quote: false
            tests:
              - not_null
          - name: region
            description: Region in which user session occurred, based on IP.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: city
            description: City in which user session occurred, based on IP.
            meta:
              type: string
              metabase.semantic_type: type/City
            quote: false
            tests:
              - not_null
          - name: IP
            description: >-
              The IP address for the session, which is used for determining
              geolocation.
            meta:
              type: string
            quote: true
            tests:
              - not_null
          - name: referrer
            description: >-
              URL that linked to your site and initiated the session.If the user
              navigated directly to your site, or referral headers were
              stripped, then this value will appear as NULL downstream and as
              \`direct\` in the UI.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: landing_page
            description: URL of the first pageview of the session.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: landing_page_query
            description: The query parameters of the first page of the user’s session.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: landing_page_hash
            description: The hash route of the first page of the user’s session.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: browser
            description: User’s browser.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: search_keyword
            description: Search term that brought the user to your site. [Deprecated]
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: utm_source
            description: >-
              GA-based utm_source tag associated with the session’s initial
              pageview.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: utm_campaign
            description: >-
              GA-based utm_campaign tag associated with the session’s initial
              pageview.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: utm_medium
            description: >-
              GA-based utm_medium tag associated with the session’s initial
              pageview.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: utm_term
            description: >-
              GA-based utm_term tag associated with the session’s initial
              pageview.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: utm_content
            description: >-
              GA-based utm_content tag associated with the session’s initial
              pageview.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: domain
            description: Domain including subdomain, e.g. blog.heap.io.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: path
            description: >-
              Portion of the current URL following your domain, e.g. /docs for
              heap.io/docs.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: hash
            description: >-
              Portion of the current URL following the hash sign, e.g. #install
              for heap.io/docs#install.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: query
            description: >-
              Query params of the page’s current URL, e.g. ?utm_id=1234 for
              heap.io?utm_id=1234.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: title
            description: Title of the current page.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: href
            description: href property of link (used for clicks on anchor tags).
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: target_text
            description: Button text of the event target.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: heap_previous_page
            description: The previous page visited in this session.
            meta:
              type: string
            quote: false
            tests:
              - not_null
"
`;

exports[`generate with dbt parses an event with Segment 1`] = `
"name: test
version: 1.0.0
config-version: 2
profile: default
model-paths:
  - models
test-paths:
  - tests
seed-paths:
  - seeds
target-path: target
clean-targets:
  - target
  - dbt_packages
"
`;

exports[`generate with dbt parses an event with Segment 2`] = `
"default:
  target: dev
  outputs:
    dev:
      type: bigquery
      method: oauth
      project: testProject
      dataset: testDataset
      timeout_seconds: 300
"
`;

exports[`generate with dbt parses an event with Segment 3`] = `
"version: 2
sources:
  - schema: testProject
    database: testDataset
    name: testSink
    description: Loaded from Syft
    tables:
      - name: test_event
        columns:
          - name: backgrounded
            meta:
              type: boolean
            quote: false
            tests:
              - not_null
          - name: _id
            description: The unique ID of the track call itself.
            meta:
              type: number
              metabase.semantic_type: type/PK
            quote: false
            tests:
              - not_null
          - name: anonymous_id
            description: The anonymous ID of the user.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: context_ip
            description: >-
              The IP address of the client. Non-user-related context fields sent
              with each track call.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: context_library_name
            description: >-
              The Logging library name. Non-user-related context fields sent
              with each track call.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: context_library_version
            description: >-
              The Logging library version. Non-user-related context fields sent
              with each track call.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: context_page_path
            description: The path of the page on which the event was logged.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: context_page_title
            description: The title of the page on which the event was logged.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: context_page_url
            description: The full url of the page on which the event was logged.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: context_locale
            description: The browsers locale used when the event was logged.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: context_user_agent
            description: The browsers user-agent string.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: event
            description: The slug of the event name, so you can join the tracks table.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: event_text
            description: The name of the event.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: loaded_at
            description: >-
              The UTC timestamp reflecting when the data was staged by the
              processor. Bigquery only.
            meta:
              type: timestamp
            quote: false
            tests:
              - not_null
          - name: original_timestamp
            description: >-
              The original timestamp set by the Segment library at the time the
              event is created. This timestamp can be affected by device clock
              skew.
            meta:
              type: timestamp
            quote: false
            tests:
              - not_null
          - name: received_at
            description: When Segment received the track call.
            meta:
              type: timestamp
            quote: false
            tests:
              - not_null
          - name: sent_at
            description: >-
              When a user triggered the track call. This timestamp can also be
              affected by device clock skew
            meta:
              type: timestamp
            quote: false
            tests:
              - not_null
          - name: timestamp
            description: The UTC-converted timestamp which is set by the Segment library
            meta:
              type: timestamp
            quote: false
            tests:
              - not_null
          - name: uuid_ts
            description: >-
              This is used to keep track of when the specific event was last
              processed by our connector, specifically for deduping and
              debugging purposes. You can generally ignore this column
            meta:
              type: timestamp
            quote: false
            tests:
              - not_null
"
`;

exports[`generate with dbt parses an event with Syft 1`] = `
"name: test
version: 1.0.0
config-version: 2
profile: default
model-paths:
  - models
test-paths:
  - tests
seed-paths:
  - seeds
target-path: target
clean-targets:
  - target
  - dbt_packages
"
`;

exports[`generate with dbt parses an event with Syft 2`] = `
"default:
  target: dev
  outputs:
    dev:
      type: bigquery
      method: oauth
      project: testProject
      dataset: testDataset
      timeout_seconds: 300
"
`;

exports[`generate with dbt parses an event with Syft 3`] = `
"version: 2
sources:
  - schema: testProject
    database: testDataset
    name: testSink
    description: Loaded from Syft
    tables:
      - name: test_event
        columns:
          - name: backgrounded
            meta:
              type: boolean
            quote: false
            tests:
              - not_null
          - name: _id
            description: The unique ID of the track call itself.
            meta:
              type: number
              metabase.semantic_type: type/PK
            quote: false
            tests:
              - not_null
          - name: anonymous_id
            description: The anonymous ID of the user.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: context_ip
            description: >-
              The IP address of the client. Non-user-related context fields sent
              with each track call.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: context_library_name
            description: >-
              The Logging library name. Non-user-related context fields sent
              with each track call.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: context_library_version
            description: >-
              The Logging library version. Non-user-related context fields sent
              with each track call.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: context_page_path
            description: The path of the page on which the event was logged.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: context_page_title
            description: The title of the page on which the event was logged.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: context_page_url
            description: The full url of the page on which the event was logged.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: context_locale
            description: The browsers locale used when the event was logged.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: context_user_agent
            description: The browsers user-agent string.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: event
            description: The slug of the event name, so you can join the tracks table.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: event_text
            description: The name of the event.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: loaded_at
            description: >-
              The UTC timestamp reflecting when the data was staged by the
              processor. Bigquery only.
            meta:
              type: timestamp
            quote: false
            tests:
              - not_null
          - name: original_timestamp
            description: >-
              The original timestamp set by the Segment library at the time the
              event is created. This timestamp can be affected by device clock
              skew.
            meta:
              type: timestamp
            quote: false
            tests:
              - not_null
          - name: received_at
            description: When Segment received the track call.
            meta:
              type: timestamp
            quote: false
            tests:
              - not_null
          - name: sent_at
            description: >-
              When a user triggered the track call. This timestamp can also be
              affected by device clock skew
            meta:
              type: timestamp
            quote: false
            tests:
              - not_null
          - name: timestamp
            description: The UTC-converted timestamp which is set by the Segment library
            meta:
              type: timestamp
            quote: false
            tests:
              - not_null
          - name: uuid_ts
            description: >-
              This is used to keep track of when the specific event was last
              processed by our connector, specifically for deduping and
              debugging purposes. You can generally ignore this column
            meta:
              type: timestamp
            quote: false
            tests:
              - not_null
          - name: event_id
            description: Unique ID of the associated event, randomly generated by Heap.
            meta:
              type: number
            quote: false
            tests:
              - not_null
          - name: user_id
            description: Unique ID of the associated user, randomly generated by Heap.
            meta:
              type: number
            quote: false
            tests:
              - not_null
          - name: session_id
            description: Unique ID of the associated session, randomly generated by Heap.
            meta:
              type: number
            quote: false
            tests:
              - not_null
          - name: time
            description: Timestamp without timezone of when the event happened.
            meta:
              type: timestamp
            quote: false
            tests:
              - not_null
          - name: type
            description: >-
              For web autocaptured events, can be any of view page, click,
              submit, change, with push state events registered as view page
              events. For iOS auto-tracked events, can be touch, edit field, or
              a gesture recognizer you’ve defined. For custom events, this will
              be the custom event name.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: library
            description: >-
              Version of Heap library on which event occurred. Can be one of
              “web”, “iOS”, or “server”.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: platform
            description: User’s operating system.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: device_type
            description: Device type, which can be one of “Mobile”, “Tablet”, or “Desktop”.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: country
            description: Country in which user session occurred, based on IP.
            meta:
              type: string
              metabase.semantic_type: type/Country
            quote: false
            tests:
              - not_null
          - name: region
            description: Region in which user session occurred, based on IP.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: city
            description: City in which user session occurred, based on IP.
            meta:
              type: string
              metabase.semantic_type: type/City
            quote: false
            tests:
              - not_null
          - name: IP
            description: >-
              The IP address for the session, which is used for determining
              geolocation.
            meta:
              type: string
            quote: true
            tests:
              - not_null
          - name: referrer
            description: >-
              URL that linked to your site and initiated the session.If the user
              navigated directly to your site, or referral headers were
              stripped, then this value will appear as NULL downstream and as
              \`direct\` in the UI.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: landing_page
            description: URL of the first pageview of the session.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: landing_page_query
            description: The query parameters of the first page of the user’s session.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: landing_page_hash
            description: The hash route of the first page of the user’s session.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: browser
            description: User’s browser.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: search_keyword
            description: Search term that brought the user to your site. [Deprecated]
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: utm_source
            description: >-
              GA-based utm_source tag associated with the session’s initial
              pageview.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: utm_campaign
            description: >-
              GA-based utm_campaign tag associated with the session’s initial
              pageview.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: utm_medium
            description: >-
              GA-based utm_medium tag associated with the session’s initial
              pageview.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: utm_term
            description: >-
              GA-based utm_term tag associated with the session’s initial
              pageview.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: utm_content
            description: >-
              GA-based utm_content tag associated with the session’s initial
              pageview.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: domain
            description: Domain including subdomain, e.g. blog.heap.io.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: path
            description: >-
              Portion of the current URL following your domain, e.g. /docs for
              heap.io/docs.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: hash
            description: >-
              Portion of the current URL following the hash sign, e.g. #install
              for heap.io/docs#install.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: query
            description: >-
              Query params of the page’s current URL, e.g. ?utm_id=1234 for
              heap.io?utm_id=1234.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: title
            description: Title of the current page.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: href
            description: href property of link (used for clicks on anchor tags).
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: target_text
            description: Button text of the event target.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: heap_previous_page
            description: The previous page visited in this session.
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: receivedAt
            description: When Syft received the track call.
            meta:
              type: timestamp
              metabase.semantic_type: type/CreationTimestamp
            quote: true
            tests:
              - not_null
          - name: eventType
            description: The slug of the event name, so you can join the tracks table.
            meta:
              type: string
              metabase.semantic_type: type/Category
            quote: true
            tests:
              - not_null
          - name: eventName
            description: The name of the event.
            meta:
              type: string
              metabase.semantic_type: type/Category
            quote: true
            tests:
              - not_null
          - name: ip
            description: N/A
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: userAgent
            description: N/A
            meta:
              type: string
            quote: true
            tests:
              - not_null
          - name: referer
            description: N/A
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: host
            description: N/A
            meta:
              type: string
            quote: false
            tests:
              - not_null
          - name: origin
            description: N/A
            meta:
              type: string
            quote: false
            tests:
              - not_null
"
`;
